### PR TITLE
[google_sign_in] Add doc for iOS auth with SERVER_CLIENT_ID

### PR DESCRIPTION
**Pull request by Milvintsiss**
Add documentation for iOS auth with SERVER_CLIENT_ID.

After a long time struggling with a non-working iOS configuration but Android working well, I figured out that firebase was not adding the SERVER_CLIENT_ID field in the Google-info-service.plist file genereated. This can be really long to troubleshoot for someone like me who don't know really much about GoogleAuth with flutter plugin.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/google_sign_in/google_sign_in/README.md [hunk 1]
```diff
@@ -61,6 +61,20 @@ This plugin requires iOS 9.0 or higher.
 <!-- End of the Google Sign-in Section -->
 ```
 
+### iOS Authenticate with backend server
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in/README.md:64]: Is there some documentation you can point us to for where this information comes from? Without context it's pretty hard to review this.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/google_sign_in/google_sign_in/README.md (hunk 1 above)
Comment (Milvintsiss) [reply packages/google_sign_in/google_sign_in/README.md:64]: No, I did'nt found any, it's why I'm doing this PR. This has been introduced in v4.5.0.

Here a comment mentioning it: https://github.com/flutter/flutter/issues/57945#issuecomment-758269872

The PR introducing it: https://github.com/flutter/plugins/pull/2116

The app example config: https://github.com/flutter/plugins/blob/f6495091d47070ec7e1aa456b236a04b674fc378/packages/google_sign_in/google_sign_in/example/ios/Runner/GoogleService-Info.plist#L41

This is the official doc for native iOS apps: https://developers.google.com/identity/sign-in/ios/offline-access

Comment (stuartmorgan) [reply packages/google_sign_in/google_sign_in/README.md:64]: @cyanglaz It looks like you reviewed the PR that added that example to the plist; could you review these instructions?

File: packages/google_sign_in/google_sign_in/README.md [hunk 2]
```diff
@@ -61,6 +61,20 @@ This plugin requires iOS 9.0 or higher.
...
+
+If you use Google Sign-In with an app that communicates with a backend server,
+you might need to identify the currently signed-in user on the server.
+To check the integrity of the token on your server you need to add an extra
+`aud` field in the `idToken`.
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/README.md:69]: What is the `aud` field in the `idToken` here?

File: packages/google_sign_in/google_sign_in/README.md [hunk 3]
```diff
@@ -61,6 +61,20 @@ This plugin requires iOS 9.0 or higher.
 <!-- End of the Google Sign-in Section -->
 ```
 
+### iOS Authenticate with backend server
+
+If you use Google Sign-In with an app that communicates with a backend server,
```
Comment (cyanglaz) [inline packages/google_sign_in/google_sign_in/README.md:66]: Please add a link to the original doc for the usage of `SERVER_CLIENT_ID`. Maybe this: https://developers.google.com/identity/sign-in/ios/backend-auth#objective-c

**Review by cyanglaz — COMMENTED** [review]
@Milvintsiss Thanks for sending this PR. Although this documentation is good to have, we probably also want to update the code to make the clientId programmatically configurable from a dart API. I think this [proposal](https://github.com/flutter/flutter/issues/57945#issuecomment-633954514) here is valid.

**Comment by Milvintsiss** [comment]
> @Milvintsiss Thanks for sending this PR. ~Although this documentation is good to have, we probably also~ ~want to update the code to make the clientId programmatically configurable from a dart API. I think this~ ~[proposal](https://github.com/flutter/flutter/issues/57945#issuecomment-633954514) here is valid.~
> 
> We actually have already implemented a way to dynamically set up the clientId, which should be petty easily discoverable. :https://github.com/flutter/plugins/blob/main/packages/google_sign_in/google_sign_in/lib/google_sign_in.dart#L186
> 
> If we are adding the plist approach for clientId in README, then we should add the programmatic approach too so users can choose the way they want to do it

As per [documentation](https://github.com/flutter/plugins/blob/3daf0b7f14c5db845d4f88dd3d07293bba6cc94b/packages/google_sign_in/google_sign_in/lib/google_sign_in.dart#L232) this is only for web. Currently there is no way to programmatically configure this.

**Comment by cyanglaz** [comment]
> `serverClientId` has currently no way to be programmatically configured, but I can implement it if needed. Maybe this deserves another PR?

I see, I got confused with these 2 IDs. Yes I don't see why it should not be programmatically configurable.
And we should update the documentation about saying the `clientId` only supports web as well.

**Comment by stuartmorgan** [comment]
@cyanglaz Can you do the primary review here since you've been looking at this the most?

**Comment by cyanglaz** [comment]
@Milvintsiss There is a PR trying to add the serverClientId https://github.com/flutter/plugins/pull/5250. It is currently being worked on. Maybe we can cooperate the new changes into this doc change after the above PR lands?

**Comment by Milvintsiss** [comment]
@cyanglaz Yes it would be great :+1:

**Comment by Hixie** [comment]
@cyanglaz @Milvintsiss looks like #5250 has landed; what are the next steps on this PR?

**Comment by Milvintsiss** [comment]
@Hixie I think we can close this one, the author has already added [documentation in the README](https://github.com/flutter/plugins/pull/5250/files#diff-bf92ca71f8218dd127d2ce164f12886d20fe34377a6267210f6cd2cd01739c31).

Maybe we can add a note about configuring the `SERVER_CLIENT_ID` in the `GoogleService-Info.plist`, but as the recommended way to configure the server client ID is to configure it programmatically we probably don't need to document this?

**Comment by cyanglaz** [comment]
> Maybe we can add a note about configuring the SERVER_CLIENT_ID in the GoogleService-Info.plist

Yes, we should also add a document for adding `SERVER_CLIENT_ID` in `GoogleService-Info.plist`. Would you like to rebase main to your PR and work on it? :)

**Comment by Milvintsiss** [comment]
Hey @cyanglaz can I have a review on this? :)

**Comment by stuartmorgan** [comment]
@cyanglaz Ping on this review from triage.

**Comment by cliuff** [comment]
Why is this issue still open? Changes regarding the doc configuring SERVER_CLIENT_ID in GoogleService-Info.plist is not yet merged. Stumbled upon this **ISSUE** today and found that setting serverClientId in dart code alone **does not work** at all. Finally got it working by adding SERVER_CLIENT_ID in GoogleService-Info.plist.

**Review by cyanglaz — APPROVED** [review]
LGTM

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Pull request merged**